### PR TITLE
NAS-105227 / 12.0 / FREENAS-MINI-3.0-E enclosures

### PR DESCRIPTION
**Pull request by themylogin**

### CODE REVIEW

File: src/freenas/usr/local/lib/middlewared_truenas/plugins/enclosure_/map.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+
+class EnclosureService(Service):
+    @private
+    async def map_enclosures(self, enclosures):
+        info = await self.middleware.call("system.info")
+        for product_mapping in MAPPINGS:
+            if product_mapping.product_re.match(info["system_product"]):
+                for version_mapping in product_mapping.mappings:
+                    if version_mapping.version_re.match(info["system_product_version"]):
```
Comment (william-gr) [inline src/freenas/usr/local/lib/middlewared_truenas/plugins/enclosure_/map.py]: Is there any hardware you're seeing that will need the version?

File: src/freenas/usr/local/lib/middlewared_truenas/plugins/enclosure_/map.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+from middlewared.service import Service, private
+
+ProductMapping = namedtuple("ProductMapping", ["product_re", "mappings"])
+VersionMapping = namedtuple("VersionMapping", ["version_re", "slots"])
+MappingSlot = namedtuple("MappingSlot", ["num", "slot", "identify"])
+
+
+MAPPINGS = [
+    ProductMapping(re.compile(r"To Be Filled By O.E.M."), [
```
Comment (william-gr) [inline src/freenas/usr/local/lib/middlewared_truenas/plugins/enclosure_/map.py]: Why this?

Comment (themylogin) [reply src/freenas/usr/local/lib/middlewared_truenas/plugins/enclosure_/map.py]: That's what was needed to test on test06, do you want me to remove that?

File: src/freenas/usr/local/lib/middlewared_truenas/plugins/enclosure_/map.py (hunk 1 above)
Comment (themylogin) [reply src/freenas/usr/local/lib/middlewared_truenas/plugins/enclosure_/map.py]: NEP says that mapping can vary from version to version

File: src/freenas/usr/local/lib/middlewared_truenas/plugins/enclosure_/map.py (hunk 2 above)
Comment (william-gr) [reply src/freenas/usr/local/lib/middlewared_truenas/plugins/enclosure_/map.py]: Yes, I was thinking just was just be there while testing in the box, not actually commit it to the repo. Or wait until we have a final say of what the system-productname will be

File: src/freenas/usr/local/lib/middlewared_truenas/plugins/enclosure_/map.py (hunk 1 above)
Comment (william-gr) [reply src/freenas/usr/local/lib/middlewared_truenas/plugins/enclosure_/map.py]: Can you remind me which section?

Comment (william-gr) [reply src/freenas/usr/local/lib/middlewared_truenas/plugins/enclosure_/map.py]: Ah, nevermind. its there just for future use and not actually used yet.

**Pull request merged**